src/test/groovy/com/stormpath/tck/register/RegisterIT.groovy

```diff
@@ -0,0 +1,576 @@
+
+        // Todo: CSRF support
+
+        Response response =
+            given()
+                .accept(ContentType.HTML)
+            .when()
+                .post(registerRoute)
+            .then()
+                .statusCode(200)
+                .contentType(ContentType.HTML)
+            .extract()
+                .response()
+
+        XmlPath doc = getHtmlDoc(response)
+
+        Node warning = findTagWithAttribute(doc.getNodeChildren("html.body"), "div", "class", "alert-danger")
+        assertThat(warning.toString(), not(isEmptyOrNullString()))
+    }
+
+    /**
+     * Re-render form with error if required fields are missing
+     * @see <a href="https://github.com/stormpath/stormpath-framework-tck/issues/188">#188</a>
+     * @throws Exception
+     */
+    @Test
+    public void displaysErrorIfEmailIsMissing() throws Exception {
+
+        // todo: work with CSRF
+
+        Response response =
+            given()
+                .accept(ContentType.HTML)
+                .formParam("email", accountEmail)
+                .formParam("password", "")
+            .when()
+                .post(registerRoute)
+            .then()
+                .statusCode(200)
+                .contentType(ContentType.HTML)
+            .extract()
+                .response()
+
+        XmlPath doc = getHtmlDoc(response)
+
+        Node warning = findTagWithAttribute(doc.getNodeChildren("html.body"), "div", "class", "alert-danger")
+        assertThat(warning.toString(), not(isEmptyOrNullString()))
+    }
+
+    /**
+     * Re-render form with error if required fields are missing
+     * @see <a href="https://github.com/stormpath/stormpath-framework-tck/issues/188">#188</a>
+     * @throws Exception
+     */
+    @Test
+    public void displaysErrorIfRequiredFieldIsMissing() throws Exception {
+
+        // todo: work with CSRF
+
+        Response response =
+            given()
+                .accept(ContentType.HTML)
+                .formParam("email", accountEmail)
+                .formParam("password", accountPassword)
+                .formParam("surname", accountSurname)
+                // givenName is required per the default configuration
+            .when()
+                .post(registerRoute)
+            .then()
+                .statusCode(200)
+                .contentType(ContentType.HTML)
+            .extract()
+                .response()
+
+        XmlPath doc = getHtmlDoc(response)
+
+        Node warning = findTagWithAttribute(doc.getNodeChildren("html.body"), "div", "class", "alert-danger")
+        assertThat(warning.toString(), not(isEmptyOrNullString()))
+    }
+
+    /**
+     * Re-render form with error if undefined custom field is present
+     * @see <a href="https://github.com/stormpath/stormpath-framework-tck/issues/194">#194</a>
+     * @throws Exception
+     */
+    @Test
+    public void displaysErrorForUndefinedCustomField() throws Exception {
```

Comment:
It might be overkill to test the validation logic twice... It might be good just to test validation logic for the JSON endpoint and then test that the HTML form displays errors (on a similar error). 

Actually, same with the form fields (instead of iterating through the list, etc); just check the JSON endpoint